### PR TITLE
Add & remove a key in key_auth

### PR DESCRIPTION
**Pull request by bonustrack**
I have some trouble making this working and interested to have your feedback.

I created 2 helpers functions: `addKeyAuth` and `removeKeyAuth` for adding and removing a key inside user account `key_auths` array. It's using the operation `account_update` to do this change (https://github.com/steemit/steem-js/blob/d81a9a3a0f22cabc9a81fd8843801fdf980ecf08/src/broadcast/helpers.js#L109-L118) .

We have already the helpers functions `addAccountAuth` and `removeAccountAuth` that working well, but for some reason when i try to add a key instead of an account i got the error "Missing Active Authority". I've verified the private active WIF and account_update operation payload and they are valid.

Any idea what is wrong?

### CODE REVIEW

**Comment by bonustrack** [comment]
Created an issue on steem repo https://github.com/steemit/steem/issues/1272

**Comment by roadscape** [comment]
From abit in above issue:
> Perhaps a serialization issue. Accounts and keys should be sorted correctly. Currently keys are converted to addresses before sorting.

**Comment by bonustrack** [comment]
Awesome it's finally working with Abit suggestion :)

File: src/broadcast/helpers.js [hunk 1]
```diff
@@ -57,14 +57,94 @@ exports = module.exports = steemBroadcast => {
...
+
+  steemBroadcast.addKeyAuth = (
+    activeWif,
+    username,
+    authorizedKey,
+    role = 'posting',
+    weight = 1,
+    cb
+  ) => {
+    api.getAccountsAsync([username]).then(([userAccount]) => {
```
Comment (yamadapc) [inline src/broadcast/helpers.js]: If this fails the callback won't be called, you want to use one of either promises or callbacks all the way

File: src/broadcast/helpers.js [hunk 2]
```diff
@@ -57,14 +57,94 @@ exports = module.exports = steemBroadcast => {
...
+  };
+
+  steemBroadcast.removeKeyAuth = (
+    activeWif,
+    username,
+    authorizedKey,
+    role = 'posting',
+    cb
+  ) => {
+    api.getAccountsAsync([username]).then(([userAccount]) => {
```
Comment (yamadapc) [inline src/broadcast/helpers.js]: Same as above

File: src/broadcast/helpers.js [hunk 3]
```diff
@@ -57,14 +57,94 @@ exports = module.exports = steemBroadcast => {
...
+      const totalAuthorizedKey = updatedAuthority.key_auths.length;
+      for (let i = 0; i < totalAuthorizedKey; i++) {
+        const user = updatedAuthority.key_auths[i];
+        if (user[0] === authorizedKey) {
+          updatedAuthority.key_auths.splice(i, 1);
+          break;
+        }
+      }
+
+      // key not exist in authorized list
```
Comment (yamadapc) [inline src/broadcast/helpers.js]: This section is slightly confusing to me, too many variables, but all good

File: src/broadcast/helpers.js [hunk 4]
```diff
@@ -57,14 +61,98 @@ exports = module.exports = steemBroadcast => {
...
+    activeWif,
+    username,
+    authorizedKey,
+    role = 'posting',
+    weight = 1,
+    cb
+  ) => {
+    api.getAccounts([username], (err, [userAccount]) => {
+      if (err) { return cb(err, null); }
+      if (!userAccount) { return cb('Invalid account name', null); }
```
Comment (yamadapc) [inline src/broadcast/helpers.js]: Hm, the first argument should always be an Error instance

File: src/broadcast/helpers.js [hunk 5]
```diff
@@ -1,31 +1,32 @@
...
+import api from '../api';
 
 exports = module.exports = steemBroadcast => {
   steemBroadcast.addAccountAuth = (
     activeWif,
     username,
     authorizedUsername,
-    role = "posting",
+    role = 'posting',
+    weight = 1,
```
Comment (sneak) [inline src/broadcast/helpers.js]: what if the account has a different threshold?

Comment (bonustrack) [inline src/broadcast/helpers.js]: If the account has a `weight_threshold` of 2 then the newly added key will not be able to broadcast alone if the weight is not set to 2. I can change the default weight to use the same value as `weight_threshold`.

File: examples/add-key-auth.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+//const pubActiveWif = steem.auth.wifToPublic(privActiveWif);
+//console.log(pubActiveWif);
+
+/* Add posting key auth */
+steem.broadcast.addKeyAuth(
+  privActiveWif,
+  username,
+  'STM88CPfhCmeEzCnvC1Cjc3DNd1DTjkMcmihih8SSxmm4LBqRq5Y9', // @fabien public posting WIF
+  'posting',
+  1,
```
Comment (sneak) [inline examples/add-key-auth.js]: i would love for these arguments to be named instead of positional

**Comment by bonustrack** [comment]
@sneak i've changed the arguments to be named instead of positional and now if weight is not specified it take the weight_thresold value as default weight.

**Comment by sneak** [comment]
awesome. in general i prefer named arguments vs positional for anything that takes more than 1-2 params. using names like signingKey instead of activeWif shows which key is which, too.

it makes it possible to audit calling code for correctness without knowing the api being called into.

**Comment by bonustrack** [comment]
Yes it make sense, on the rest of the library both solutions (named and positional) are possible example:
```
steem.broadcast.transferWith(key, { from, to, amount, memo }, cb) // named
steem.broadcast.transfer(key, from, to, amount, memo, cb) // positional
```
I've renamed `activeWif` to `signingKey` also.

**Pull request merged**